File: MinimumDepthOfBinaryTree.java
/**
 * Created by yhf on 15/4/15.
 */

/*
Given a binary tree, find its minimum depth.

The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.

       5
      / \
     4   8
    /   / \
  11  13   4
 /  \       \
7    2       1
*/

public class MinimumDepthOfBinaryTree {
    public static int minDepth(TreeNode root) {
        // 注意，是到叶子节点
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            return 1;
        }
        if (root.left == null) {
            return minDepth(root.right) + 1;
        } else if (root.right == null) {
            return minDepth(root.left) + 1;
        } else {
            return Math.min(minDepth(root.left), minDepth(root.right)) + 1;
        }
    }

    public static void main(String[] args) {
        TreeNode n1 = new TreeNode(5);
        TreeNode n2 = new TreeNode(4);
        TreeNode n3 = new TreeNode(8);
        TreeNode n4 = new TreeNode(11);
        TreeNode n5 = new TreeNode(13);
        TreeNode n6 = new TreeNode(4);
        TreeNode n7 = new TreeNode(7);
        TreeNode n8 = new TreeNode(2);
        TreeNode n9 = new TreeNode(1);
        n1.left = n2;
        n1.right = n3;
        n2.left = n4;
        n3.left = n5;
        n3.right = n6;
        n4.left = n7;
        n4.right = n8;
        n6.right = n9;
        System.out.println(minDepth(n1));
    }
}

File: MinimumDepthOfBinaryTree.py
"""
Given a binary tree, find its minimum depth.

The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
"""
from TreeNode import TreeNode


class Solution:
    # @param {TreeNode} root
    # @return {integer}
    def minDepth(self, root):
        if not root:
            return 0

        self.__min_depth = 2 ** 31

        def __minDepth(root, depth):
            if not root.left and not root.right:
                self.__min_depth = min(depth, self.__min_depth)
            if root.left:
                __minDepth(root.left, depth + 1)
            if root.right:
                __minDepth(root.right, depth + 1)

        __minDepth(root, 1)
        return self.__min_depth


if __name__ == '__main__':
    n1 = TreeNode(1)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n1.left = n2
    n1.right = n3
    print Solution().minDepth(n1)